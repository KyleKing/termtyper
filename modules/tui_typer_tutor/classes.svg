<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN"
 "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<!-- Generated by graphviz version 8.1.0 (20230707.0739)
 -->
<!-- Title: classes Pages: 1 -->
<svg width="1659pt" height="507pt"
 viewBox="0.00 0.00 1658.62 506.50" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
<g id="graph0" class="graph" transform="scale(1 1) rotate(0) translate(4 502.5)">
<title>classes</title>
<polygon fill="white" stroke="none" points="-4,4 -4,-502.5 1654.62,-502.5 1654.62,4 -4,4"/>
<!-- tui_typer_tutor.core.typing.AtEndOfExpectedError -->
<g id="node1" class="node">
<title>tui_typer_tutor.core.typing.AtEndOfExpectedError</title>
<polygon fill="none" stroke="black" points="0,-78.75 0,-127.75 144.25,-127.75 144.25,-78.75 0,-78.75"/>
<text text-anchor="start" x="8" y="-110.45" font-family="Times,serif" font-size="14.00" fill="red">AtEndOfExpectedError</text>
<polyline fill="none" stroke="black" points="0,-103.25 144.25,-103.25"/>
<polyline fill="none" stroke="black" points="0,-103.25 144.25,-103.25"/>
<text text-anchor="start" x="70.25" y="-85.95" font-family="Times,serif" font-size="14.00" fill="red"> </text>
</g>
<!-- tui_typer_tutor.core.config.Config -->
<g id="node2" class="node">
<title>tui_typer_tutor.core.config.Config</title>
<polygon fill="none" stroke="black" points="162.12,-50 162.12,-156.5 262.12,-156.5 262.12,-50 162.12,-50"/>
<text text-anchor="start" x="193.38" y="-139.2" font-family="Times,serif" font-size="14.00">Config</text>
<polyline fill="none" stroke="black" points="162.12,-132 262.12,-132"/>
<text text-anchor="start" x="170.12" y="-114.7" font-family="Times,serif" font-size="14.00">model_config</text>
<text text-anchor="start" x="170.12" y="-98.2" font-family="Times,serif" font-size="14.00">seed_file : Path</text>
<text text-anchor="start" x="170.12" y="-81.7" font-family="Times,serif" font-size="14.00">seed_file : Path</text>
<polyline fill="none" stroke="black" points="162.12,-74.5 262.12,-74.5"/>
<text text-anchor="start" x="210.25" y="-57.2" font-family="Times,serif" font-size="14.00"> </text>
</g>
<!-- tui_typer_tutor.core.typing.ExpectedKey -->
<g id="node3" class="node">
<title>tui_typer_tutor.core.typing.ExpectedKey</title>
<polygon fill="none" stroke="black" points="357.75,-333.75 357.75,-423.75 446.5,-423.75 446.5,-333.75 357.75,-333.75"/>
<text text-anchor="start" x="365.75" y="-406.45" font-family="Times,serif" font-size="14.00">ExpectedKey</text>
<polyline fill="none" stroke="black" points="357.75,-399.25 446.5,-399.25"/>
<text text-anchor="start" x="371.38" y="-381.95" font-family="Times,serif" font-size="14.00">text</text>
<text text-anchor="start" x="371.38" y="-365.45" font-family="Times,serif" font-size="14.00">textual : str</text>
<polyline fill="none" stroke="black" points="357.75,-358.25 446.5,-358.25"/>
<text text-anchor="start" x="400.25" y="-340.95" font-family="Times,serif" font-size="14.00"> </text>
</g>
<!-- tui_typer_tutor.screens.help.Help -->
<g id="node4" class="node">
<title>tui_typer_tutor.screens.help.Help</title>
<polygon fill="none" stroke="black" points="542.25,-50 542.25,-156.5 760,-156.5 760,-50 542.25,-50"/>
<text text-anchor="start" x="638" y="-139.2" font-family="Times,serif" font-size="14.00">Help</text>
<polyline fill="none" stroke="black" points="542.25,-132 760,-132"/>
<text text-anchor="start" x="550.25" y="-114.7" font-family="Times,serif" font-size="14.00">BINDINGS : ClassVar[list[Binding]]</text>
<text text-anchor="start" x="550.25" y="-98.2" font-family="Times,serif" font-size="14.00">DEFAULT_CSS : ClassVar[str]</text>
<polyline fill="none" stroke="black" points="542.25,-91 760,-91"/>
<text text-anchor="start" x="575.38" y="-73.7" font-family="Times,serif" font-size="14.00">action_close(): None</text>
<text text-anchor="start" x="575.38" y="-57.2" font-family="Times,serif" font-size="14.00">compose(): ComposeResult</text>
</g>
<!-- tui_typer_tutor.core.typing.Keys -->
<g id="node5" class="node">
<title>tui_typer_tutor.core.typing.Keys</title>
<polygon fill="none" stroke="black" points="778.38,-25.25 778.38,-181.25 951.88,-181.25 951.88,-25.25 778.38,-25.25"/>
<text text-anchor="start" x="851.25" y="-163.95" font-family="Times,serif" font-size="14.00">Keys</text>
<polyline fill="none" stroke="black" points="778.38,-156.75 951.88,-156.75"/>
<text text-anchor="start" x="786.38" y="-139.45" font-family="Times,serif" font-size="14.00">expected : list[ExpectedKey]</text>
<text text-anchor="start" x="786.38" y="-122.95" font-family="Times,serif" font-size="14.00">last_was_delete</text>
<text text-anchor="start" x="786.38" y="-106.45" font-family="Times,serif" font-size="14.00">last_was_delete : bool</text>
<text text-anchor="start" x="786.38" y="-89.95" font-family="Times,serif" font-size="14.00">typed</text>
<text text-anchor="start" x="786.38" y="-73.45" font-family="Times,serif" font-size="14.00">typed : list[TypedKey]</text>
<text text-anchor="start" x="786.38" y="-56.95" font-family="Times,serif" font-size="14.00">typed_all : list[TypedKey]</text>
<polyline fill="none" stroke="black" points="778.38,-49.75 951.88,-49.75"/>
<text text-anchor="start" x="829.12" y="-32.45" font-family="Times,serif" font-size="14.00">store(): None</text>
</g>
<!-- tui_typer_tutor.screens.main.Main -->
<g id="node6" class="node">
<title>tui_typer_tutor.screens.main.Main</title>
<polygon fill="none" stroke="black" points="883.25,-259.5 883.25,-498 1101,-498 1101,-259.5 883.25,-259.5"/>
<text text-anchor="start" x="977.5" y="-480.7" font-family="Times,serif" font-size="14.00">Main</text>
<polyline fill="none" stroke="black" points="883.25,-473.5 1101,-473.5"/>
<text text-anchor="start" x="891.25" y="-456.2" font-family="Times,serif" font-size="14.00">BINDINGS : ClassVar[list[Binding]]</text>
<text text-anchor="start" x="891.25" y="-439.7" font-family="Times,serif" font-size="14.00">DEFAULT_CSS : ClassVar[str]</text>
<text text-anchor="start" x="891.25" y="-423.2" font-family="Times,serif" font-size="14.00">keys</text>
<text text-anchor="start" x="891.25" y="-406.7" font-family="Times,serif" font-size="14.00">keys</text>
<text text-anchor="start" x="891.25" y="-390.2" font-family="Times,serif" font-size="14.00">metrics</text>
<text text-anchor="start" x="891.25" y="-373.7" font-family="Times,serif" font-size="14.00">metrics</text>
<text text-anchor="start" x="891.25" y="-357.2" font-family="Times,serif" font-size="14.00">width</text>
<text text-anchor="start" x="891.25" y="-340.7" font-family="Times,serif" font-size="14.00">width : int</text>
<polyline fill="none" stroke="black" points="883.25,-333.5 1101,-333.5"/>
<text text-anchor="start" x="910.38" y="-316.2" font-family="Times,serif" font-size="14.00">action_save_and_quit(): None</text>
<text text-anchor="start" x="910.38" y="-299.7" font-family="Times,serif" font-size="14.00">compose(): ComposeResult</text>
<text text-anchor="start" x="910.38" y="-283.2" font-family="Times,serif" font-size="14.00">on_key(event: Key): None</text>
<text text-anchor="start" x="910.38" y="-266.7" font-family="Times,serif" font-size="14.00">on_mount(): None</text>
</g>
<!-- tui_typer_tutor.core.typing.Keys&#45;&gt;tui_typer_tutor.screens.main.Main -->
<g id="edge4" class="edge">
<title>tui_typer_tutor.core.typing.Keys&#45;&gt;tui_typer_tutor.screens.main.Main</title>
<path fill="none" stroke="black" d="M887.95,-180.93C894.7,-200.81 902.66,-221.98 911.38,-241 912.43,-243.3 913.51,-245.62 914.62,-247.94"/>
<polygon fill="black" stroke="black" points="914.66,-248.01 920.9,-251.62 919.96,-258.77 913.72,-255.16 914.66,-248.01"/>
<text text-anchor="middle" x="924.5" y="-227.7" font-family="Times,serif" font-size="14.00" fill="green">keys</text>
</g>
<!-- tui_typer_tutor.core.typing.Keys&#45;&gt;tui_typer_tutor.screens.main.Main -->
<g id="edge5" class="edge">
<title>tui_typer_tutor.core.typing.Keys&#45;&gt;tui_typer_tutor.screens.main.Main</title>
<path fill="none" stroke="black" d="M913.28,-181.07C921.37,-195.26 929.33,-210.15 936.12,-224.5 939.71,-232.07 943.21,-239.93 946.6,-247.93"/>
<polygon fill="black" stroke="black" points="946.54,-247.79 952.53,-251.8 951.14,-258.87 945.14,-254.86 946.54,-247.79"/>
<text text-anchor="middle" x="956.5" y="-227.7" font-family="Times,serif" font-size="14.00" fill="green">keys</text>
</g>
<!-- tui_typer_tutor.core.metrics.SessionMetrics -->
<g id="node7" class="node">
<title>tui_typer_tutor.core.metrics.SessionMetrics</title>
<polygon fill="none" stroke="black" points="969.62,-0.5 969.62,-206 1240.62,-206 1240.62,-0.5 969.62,-0.5"/>
<text text-anchor="start" x="1063.5" y="-188.7" font-family="Times,serif" font-size="14.00">SessionMetrics</text>
<polyline fill="none" stroke="black" points="969.62,-181.5 1240.62,-181.5"/>
<text text-anchor="start" x="995.25" y="-164.2" font-family="Times,serif" font-size="14.00">filename : str</text>
<text text-anchor="start" x="995.25" y="-147.7" font-family="Times,serif" font-size="14.00">session_end : Optional[datetime | None]</text>
<text text-anchor="start" x="995.25" y="-131.2" font-family="Times,serif" font-size="14.00">session_end : datetime</text>
<text text-anchor="start" x="995.25" y="-114.7" font-family="Times,serif" font-size="14.00">session_start : datetime</text>
<text text-anchor="start" x="995.25" y="-98.2" font-family="Times,serif" font-size="14.00">typed_correct</text>
<text text-anchor="start" x="995.25" y="-81.7" font-family="Times,serif" font-size="14.00">typed_correct : int</text>
<text text-anchor="start" x="995.25" y="-65.2" font-family="Times,serif" font-size="14.00">typed_incorrect</text>
<text text-anchor="start" x="995.25" y="-48.7" font-family="Times,serif" font-size="14.00">typed_incorrect : int</text>
<polyline fill="none" stroke="black" points="969.62,-41.5 1240.62,-41.5"/>
<text text-anchor="start" x="977.62" y="-24.2" font-family="Times,serif" font-size="14.00">end_session(keys: Keys): &#39;SessionMetrics&#39;</text>
<text text-anchor="start" x="977.62" y="-7.7" font-family="Times,serif" font-size="14.00">from_filename(filename: str): &#39;SessionMetrics&#39;</text>
</g>
<!-- tui_typer_tutor.core.metrics.SessionMetrics&#45;&gt;tui_typer_tutor.screens.main.Main -->
<g id="edge2" class="edge">
<title>tui_typer_tutor.core.metrics.SessionMetrics&#45;&gt;tui_typer_tutor.screens.main.Main</title>
<path fill="none" stroke="black" d="M1046.08,-205.81C1043.08,-212.06 1040.24,-218.33 1037.62,-224.5 1034.47,-231.97 1031.44,-239.73 1028.55,-247.62"/>
<polygon fill="black" stroke="black" points="1028.59,-247.51 1030.35,-254.5 1024.57,-258.82 1022.81,-251.82 1028.59,-247.51"/>
<text text-anchor="middle" x="1057.38" y="-227.7" font-family="Times,serif" font-size="14.00" fill="green">metrics</text>
</g>
<!-- tui_typer_tutor.core.metrics.SessionMetrics&#45;&gt;tui_typer_tutor.screens.main.Main -->
<g id="edge3" class="edge">
<title>tui_typer_tutor.core.metrics.SessionMetrics&#45;&gt;tui_typer_tutor.screens.main.Main</title>
<path fill="none" stroke="black" d="M1089.4,-205.91C1086.27,-217.93 1082.55,-229.87 1078.12,-241 1077.19,-243.35 1076.22,-245.7 1075.21,-248.05"/>
<polygon fill="black" stroke="black" points="1075.21,-248.05 1076.37,-255.16 1070.25,-258.97 1069.09,-251.86 1075.21,-248.05"/>
<text text-anchor="middle" x="1103.38" y="-227.7" font-family="Times,serif" font-size="14.00" fill="green">metrics</text>
</g>
<!-- tui_typer_tutor.app.ttt.TuiTyperTutor -->
<g id="node8" class="node">
<title>tui_typer_tutor.app.ttt.TuiTyperTutor</title>
<polygon fill="none" stroke="black" points="1258.25,-50 1258.25,-156.5 1476,-156.5 1476,-50 1258.25,-50"/>
<text text-anchor="start" x="1326.62" y="-139.2" font-family="Times,serif" font-size="14.00">TuiTyperTutor</text>
<polyline fill="none" stroke="black" points="1258.25,-132 1476,-132"/>
<text text-anchor="start" x="1266.25" y="-114.7" font-family="Times,serif" font-size="14.00">BINDINGS : ClassVar[list[Binding]]</text>
<text text-anchor="start" x="1266.25" y="-98.2" font-family="Times,serif" font-size="14.00">TITLE : str</text>
<polyline fill="none" stroke="black" points="1258.25,-91 1476,-91"/>
<text text-anchor="start" x="1295.12" y="-73.7" font-family="Times,serif" font-size="14.00">action_show_help(): None</text>
<text text-anchor="start" x="1295.12" y="-57.2" font-family="Times,serif" font-size="14.00">on_mount(): None</text>
</g>
<!-- tui_typer_tutor.core.typing.TypedKey -->
<g id="node9" class="node">
<title>tui_typer_tutor.core.typing.TypedKey</title>
<polygon fill="none" stroke="black" points="280.12,-58.25 280.12,-148.25 524.12,-148.25 524.12,-58.25 280.12,-58.25"/>
<text text-anchor="start" x="373.62" y="-130.95" font-family="Times,serif" font-size="14.00">TypedKey</text>
<polyline fill="none" stroke="black" points="280.12,-123.75 524.12,-123.75"/>
<text text-anchor="start" x="288.12" y="-106.45" font-family="Times,serif" font-size="14.00">expected : Optional[ExpectedKey | None]</text>
<text text-anchor="start" x="288.12" y="-89.95" font-family="Times,serif" font-size="14.00">was_correct</text>
<polyline fill="none" stroke="black" points="280.12,-82.75 524.12,-82.75"/>
<text text-anchor="start" x="400.25" y="-65.45" font-family="Times,serif" font-size="14.00"> </text>
</g>
<!-- tui_typer_tutor.core.typing.TypedKey&#45;&gt;tui_typer_tutor.core.typing.ExpectedKey -->
<g id="edge1" class="edge">
<title>tui_typer_tutor.core.typing.TypedKey&#45;&gt;tui_typer_tutor.core.typing.ExpectedKey</title>
<path fill="none" stroke="black" d="M402.12,-148.19C402.12,-195.74 402.12,-271.64 402.12,-323.02"/>
<polygon fill="none" stroke="black" points="398.63,-322.92 402.12,-332.92 405.63,-322.92 398.63,-322.92"/>
</g>
<!-- tui_typer_tutor._BeartypeModes -->
<g id="node10" class="node">
<title>tui_typer_tutor._BeartypeModes</title>
<polygon fill="none" stroke="black" points="1493.62,-66.5 1493.62,-140 1650.62,-140 1650.62,-66.5 1493.62,-66.5"/>
<text text-anchor="start" x="1525.62" y="-122.7" font-family="Times,serif" font-size="14.00">_BeartypeModes</text>
<polyline fill="none" stroke="black" points="1493.62,-115.5 1650.62,-115.5"/>
<text text-anchor="start" x="1557.12" y="-98.2" font-family="Times,serif" font-size="14.00">name</text>
<polyline fill="none" stroke="black" points="1493.62,-91 1650.62,-91"/>
<text text-anchor="start" x="1501.62" y="-73.7" font-family="Times,serif" font-size="14.00">from_environment(): Self</text>
</g>
</g>
</svg>
