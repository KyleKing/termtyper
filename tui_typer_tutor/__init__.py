"""tui_typer_tutor."""

__version__ = '0.0.1'
__pkg_name__ = 'tui_typer_tutor'

# ====== Above is the recommended code from calcipy_template and may be updated on new releases ======

APP_NAME = 'tui-typer-tutor'
