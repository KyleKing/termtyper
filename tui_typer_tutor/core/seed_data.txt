def __init__(self, *args, **kwargs):<Return>super().__init__()<Return>self._counter = 0<Esc>
YY
ciw
vt"p
Vkky
<Space>ld
'
!
"
#
$
%
&
(
)
*
+
,
-
.
/
:
;
<Bar>
<Bslash>
<Return>
<Tab>
=
?
@
[
]
^
_
{
}
~
}
