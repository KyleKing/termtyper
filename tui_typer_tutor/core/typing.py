"""Typing Logic."""

import bisect

from beartype import beartype
from pydantic import BaseModel, Field

from ..constants import VIM_TO_TEXTUAL


class ExpectedKey(BaseModel):
    """Expected Key."""

    raw: str
    """Textual Key Name."""

    @property
    def text(self) -> str:
        """Displayed text."""
        return VIM_TO_TEXTUAL.inverse.get(self.raw) or self.raw

    @property
    def width(self) -> int:
        """Displayed text."""
        return len(self.text)


class TypedKey(ExpectedKey):
    """Typed Key."""

    expected: ExpectedKey
    """Store the expected key when typed and expected become out-of-sync."""

    @property
    def was_correct(self) -> bool:
        """If typed key matches expected."""
        return self.text == self.expected.text


class _KeysAccum(BaseModel):
    """Histogram-style store for identifying character widths."""

    expected: list[int] = Field(default_factory=list)
    """Accumulating count of character width per Key."""

    typed: list[int] = Field(default_factory=list)
    """Accumulating count of character width per Key."""

    @beartype
    def get_indices(self, accum: list[int], start: int, end: int) -> tuple[int, int]:
        """Return the adjusted start and end."""
        adj_start = bisect.bisect(accum, start)
        adj_end = bisect.bisect(accum, end - start + adj_start) if len(accum) > end else len(accum)
        return (adj_start, adj_end)


class Keys(BaseModel):
    """Key Model."""

    expected: list[ExpectedKey] = Field(default_factory=list)
    """The expected keys for practice."""

    typed_all: list[TypedKey] = Field(default_factory=list)
    """Append-only."""

    typed: list[TypedKey] = Field(default_factory=list)
    """Only tracks non-deleted keys."""

    last_was_delete: bool = False
    """Indicate if last operation was a delete."""

    accum: _KeysAccum = Field(default_factory=_KeysAccum)
    """Accumulating index store (private)."""

    def copy(self) -> 'Keys':
        """Ensure that the lists are deep-copied."""
        keys = super().copy()
        keys.expected = [*keys.expected]
        keys.typed_all = [*keys.typed_all]
        keys.typed = [*keys.typed]
        keys.accum.expected = [*keys.accum.expected]
        keys.accum.typed = [*keys.accum.typed]
        return keys

    @beartype
    def get_expected(self, start: int, end: int) -> list[ExpectedKey]:
        """Retrieve the next page of expected keys."""
        if not self.accum.expected:  # Lazily evaluate the expected values
            counter = 0
            for exp in self.expected:
                counter += exp.width
                self.accum.expected.append(counter)
        adj_start, adj_end = self.accum.get_indices(self.accum.expected, start, end)
        return self.expected[adj_start:adj_end]

    @beartype
    def get_typed(self, start: int, end: int) -> list[TypedKey]:
        """Retrieve the next page of typed keys."""
        adj_start, adj_end = self.accum.get_indices(self.accum.typed, start, end)
        return self.typed[adj_start:adj_end]

    @beartype
    def store(self, key: TypedKey) -> None:
        """Store a new typed key."""
        self.typed_all.append(key)
        self.last_was_delete = key.raw == 'backspace'
        if self.last_was_delete:
            if self.typed:
                self.typed = self.typed[:-1]
                self.accum.typed.pop()
        else:
            self.typed.append(key)
            counter = (self.accum.typed or [0])[-1]
            self.accum.typed.append(counter + key.width)


@beartype
def on_keypress(raw: str, keys: Keys) -> None:
    """Process a key press."""
    expected = keys.expected[len(keys.typed)]
    key = TypedKey(raw=raw, expected=expected)
    keys.store(key)
