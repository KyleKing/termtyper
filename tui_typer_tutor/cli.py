"""Extend Invoke for gh-ops."""

from types import ModuleType

from beartype import beartype
from beartype.typing import Optional
from invoke.collection import Collection
from invoke.program import Program


# PLANNED: Reduce the non-gh-ops output from Invoke
#  https://docs.pyinvoke.org/en/stable/concepts/library.html
@beartype
def start_program(  # pragma: no cover
    pkg_name: str,
    pkg_version: str,
    module: Optional[ModuleType] | None = None,
    collection: Optional[Collection] | None = None,
) -> None:
    """Run the customized Invoke Program."""
    Program(
        name=pkg_name,
        version=pkg_version,
        namespace=Collection.from_module(module) if module else collection,
    ).run()
