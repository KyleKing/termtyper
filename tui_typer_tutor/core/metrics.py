"""Manage Metrics."""

import csv
from datetime import datetime
from pathlib import Path
from zoneinfo import ZoneInfo

from beartype import beartype
from pydantic import BaseModel

from .typing import Keys


@beartype
def utcnow() -> datetime:
    """Return the current time in UTC."""
    return datetime.now(tz=ZoneInfo('UTC'))


class SessionMetrics(BaseModel):
    """Session metrics."""

    filename: str
    session_start: datetime
    session_end: datetime | None = None
    typed_correct: int = 0
    typed_incorrect: int = 0

    @classmethod
    def from_filename(cls, filename: str) -> 'SessionMetrics':  # noqa: RBT002
        """Initialize Metrics based on the filename."""
        return cls(filename=filename, session_start=utcnow())

    def end_session(self, keys: Keys) -> 'SessionMetrics':  # noqa: RBT002
        """Update the typed counters based on `Keys`."""
        self.session_end = utcnow()
        for key in keys.typed_all:
            if key.expected:
                if key.was_correct:
                    self.typed_correct += 1
                else:
                    self.typed_incorrect += 1
        return self


# PLANNED: Support Linux/Windows
CSV_PATH = Path.home() / '.config/tui-typer-tutor/metrics.csv'


@beartype
def append_csv(metrics: SessionMetrics) -> None:
    """Write metrics to the global CSV."""
    csv_columns = ['filename', 'session_start', 'session_end', 'typed_correct', 'typed_incorrect']
    if not CSV_PATH.is_file():
        CSV_PATH.parent.mkdir(exist_ok=True, parents=True)
        with CSV_PATH.open(mode='w', newline='', encoding='utf-8') as _f:
            csv.writer(_f).writerow(csv_columns)  # nosemgrep

    ser_metrics = metrics.dict()
    metrics_row = [ser_metrics[_c] for _c in csv_columns]
    with CSV_PATH.open('a', newline='', encoding='utf-8') as _f:
        csv.writer(_f).writerow(metrics_row)  # nosemgrep
