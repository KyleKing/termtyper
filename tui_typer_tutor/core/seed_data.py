"""Load Seed Data."""

import random

from beartype import beartype

from ..constants import DISPLAY_TO_TEXTUAL
from .typing import ExpectedKey


@beartype
def load_seed_data(seed_text: str) -> list[ExpectedKey]:
    """Load Seed Data in Vim format."""
    grouped_keys = [
        [ExpectedKey(textual=DISPLAY_TO_TEXTUAL[token]) for token in line]
        for line in seed_text.split('\n')
    ]
    random.shuffle(grouped_keys)  # noqa: DUO102
    return [_k for _keys in grouped_keys for _k in _keys]
