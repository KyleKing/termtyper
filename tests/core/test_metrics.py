from pathlib import Path

import arrow
import pandas as pd
import platformdirs
import pytest
from freezegun import freeze_time

from tui_typer_tutor import APP_NAME
from tui_typer_tutor.core.metrics import SessionMetrics, append_csv
from tui_typer_tutor.core.typing import on_keypress

from ..configuration import FROZEN_TIME, stubbed_user_cache_dir
from ..seed_data import lol_keys


@pytest.mark.parametrize('iterations', [1, 2])
def test_append_csv(
    iterations: int,
    snapshot,
    monkeypatch,
    fix_test_cache,  # noqa: ARG001
):
    monkeypatch.setattr(platformdirs, 'user_cache_dir', stubbed_user_cache_dir)
    csv_path = Path(stubbed_user_cache_dir(APP_NAME)) / 'metrics.csv'
    keys = lol_keys()

    all_metrics = []
    frozen_time = FROZEN_TIME
    for idx in range(1, iterations + 1):
        with freeze_time(frozen_time):
            metrics = SessionMetrics.from_filename(filename='seed_file')

        frozen_time = arrow.get(frozen_time).shift(minutes=idx).datetime
        with freeze_time(frozen_time):
            append_csv(metrics.end_session(keys))  # act
        all_metrics.append(metrics)
        on_keypress(str(idx), keys)

    df_csv = pd.read_csv(csv_path)
    assert {
        'metrics': all_metrics[-1].dict(),
        'csv': df_csv.to_dict(orient='list'),
    } == snapshot
