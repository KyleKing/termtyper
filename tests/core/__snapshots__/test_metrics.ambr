# serializer version: 1
# name: test_append_csv[1]
  dict({
    'csv': dict({
      'filename': list([
        'seed_file',
      ]),
      'session_end': list([
        '2025-01-02 14:06:06+00:00',
      ]),
      'session_start': list([
        '2025-01-02 14:05:06+00:00',
      ]),
      'typed_correct': list([
        1,
      ]),
      'typed_incorrect': list([
        0,
      ]),
    }),
    'metrics': dict({
      'filename': 'seed_file',
      'session_end': FakeDatetime(2025, 1, 2, 14, 6, 6, tzinfo=zoneinfo.ZoneInfo(key='UTC')),
      'session_start': FakeDatetime(2025, 1, 2, 14, 5, 6, tzinfo=zoneinfo.ZoneInfo(key='UTC')),
      'typed_correct': 1,
      'typed_incorrect': 0,
    }),
  })
# ---
# name: test_append_csv[2]
  dict({
    'csv': dict({
      'filename': list([
        'seed_file',
        'seed_file',
      ]),
      'session_end': list([
        '2025-01-02 14:06:06+00:00',
        '2025-01-02 14:08:06+00:00',
      ]),
      'session_start': list([
        '2025-01-02 14:05:06+00:00',
        '2025-01-02 14:06:06+00:00',
      ]),
      'typed_correct': list([
        1,
        1,
      ]),
      'typed_incorrect': list([
        0,
        1,
      ]),
    }),
    'metrics': dict({
      'filename': 'seed_file',
      'session_end': FakeDatetime(2025, 1, 2, 14, 8, 6, tzinfo=zoneinfo.ZoneInfo(key='UTC')),
      'session_start': FakeDatetime(2025, 1, 2, 14, 6, 6, tzinfo=zoneinfo.ZoneInfo(key='UTC')),
      'typed_correct': 1,
      'typed_incorrect': 1,
    }),
  })
# ---
